### PR TITLE
Sending exact asset where vulnerability found

### PR DESCRIPTION
**Pull request by oussamaessaji**

### CODE REVIEW

File: tests/agent_test.py [hunk 1]
```diff
@@ -183,7 +195,12 @@ def testTsunamiAgent_WhenServiceAssetAndTsunamiScanHasVulnerabilities_ShouldRepo
 
     mock_report_vulnerability.assert_called_once_with(entry=kb_entry,
                                                       technical_detail=f'```json\n{data}\n```',
-                                                      risk_rating=agent_report_vulnerability_mixin.RiskRating.HIGH)
+                                                      risk_rating=agent_report_vulnerability_mixin.RiskRating.HIGH,
+                                                      vulnerability_location=agent_report_vulnerability_mixin.VulnerabilityLocation(
+                                                          metadata=[agent_report_vulnerability_mixin.VulnerabilityLocationMetaData(agent_report_vulnerability_mixin.MetaDataType.PORT, '6000')],
+                                                          asset=domain_asset.DomainName(name='https://ostorlab.co')
```
Comment (deadly-panda) [inline tests/agent_test.py:201]: If you are using DomainName asset, then i expect the name to be a domainname, not a link. you add the link as part of the metadata.

Comment (deadly-panda) [inline tests/agent_test.py:201]: If you are using DomainName asset, then i expect the name to be a domain name, not a link.
You can add the link as part of the metadata.

Comment (deadly-panda) [inline tests/agent_test.py:201]: If you are using DomainName asset, then i expect the name to be a domain name, not a link.
You can add the link as part of the metadata.

File: tests/agent_test.py [hunk 2]
```diff
@@ -8,6 +8,9 @@
 
 from agent.tsunami import tsunami
 from agent import tsunami_agent as ts_agt
+from typing import List
+from ostorlab.assets import ipv4 as ipv4_asset
```
Comment (deadly-panda) [inline tests/agent_test.py:12]: Please respect the guidlines we use for the imports, standard library /n 3rd party packages /n your code in other files.

File: tests/agent_test.py [hunk 3]
```diff
@@ -268,3 +290,27 @@ def testTsunamiAgent_WhenMessageIsIpRange_ShouldCallTsunamiForAllHosts(mocker: p
...
+                    'description': 'Ostorlab is not password protected'
+                }
+            }
+        ]
+    }
+    mocker.patch('agent.tsunami.tsunami.Tsunami.scan', return_value=data)
+    tsunami_agent_no_scope.process(ip_small_range_message)
+
+    assert 'v3.report.vulnerability' in [a.selector for a in agent_mock]
+    assert ['ipv4'] in [list(a.data.get('vulnerability_location', {}).keys()) for a in agent_mock]
```
Comment (deadly-panda) [inline tests/agent_test.py:316]: Just prepare the full vuln location dict and check if it is part of the sent data.

File: tests/conftest.py [hunk 4]
```diff
@@ -40,3 +40,16 @@ def fixture_tsunami_agent_no_scope(agent_mock: List[message.Message],
...
+@pytest.fixture
+def ip_small_range_message() -> message.Message:
+    """Creates a dummy message of type v3.asset.ip.v4 with a small mask to be used by the agent for testing purposes."""
+    selector = 'v3.asset.ip.v4'
+    msg_data = {
+        'host': '42.42.42.42',
+        'mask': '31',
+        'version': 4
+    }
+    return message.Message.from_data(selector, data=msg_data)
```
Comment (deadly-panda) [inline tests/conftest.py:55]:
```suggestion
    return message.Message.from_data(selector, data=msg_data)

```

File: tests/agent_test.py [hunk 5]
```diff
@@ -133,7 +140,12 @@ def testTsunamiAgent_WhenLinkAssetAndTsunamiScanHasVulnerabilities_ShouldReportV
 
     mock_report_vulnerability.assert_called_once_with(entry=kb_entry,
                                                       technical_detail=f'```json\n{data}\n```',
-                                                      risk_rating=agent_report_vulnerability_mixin.RiskRating.HIGH)
+                                                      risk_rating=agent_report_vulnerability_mixin.RiskRating.HIGH,
+                                                      vulnerability_location=agent_report_vulnerability_mixin.VulnerabilityLocation(
+                                                          metadata=[],
+                                                          asset=domain_asset.DomainName(name='https://ostorlab.co')
```
Comment (deadly-panda) [inline tests/agent_test.py:146]: same here

File: agent/tsunami_agent.py [hunk 6]
```diff
@@ -46,6 +49,31 @@ def _check_asset_was_added(self, targets: tsunami.Target) -> bool:
                 return False
         return True
 
+    def _get_vuln_location(self, target):
```
Comment (deadly-panda) [inline agent/tsunami_agent.py:52]: add the types please

File: agent/tsunami_agent.py [hunk 7]
```diff
@@ -46,6 +49,31 @@ def _check_asset_was_added(self, targets: tsunami.Target) -> bool:
...
+        elif target.domain is not None:
+            url = urllib.parse.urlparse(target.domain)
+            if url.port is not None:
+                metadata_type = agent_report_vulnerability_mixin.MetaDataType.PORT
+                metadata_value = str(url.port)
+                metadata = [
+                    agent_report_vulnerability_mixin.VulnerabilityLocationMetaData(type=metadata_type,
+                                                                                   value=metadata_value)
+                ]
+            asset = domain_asset.DomainName(name=target.domain)
```
Comment (deadly-panda) [inline agent/tsunami_agent.py:73]: This should be the hostname, parsed_url.hostname

File: agent/tsunami/tsunami.py [hunk 8]
```diff
@@ -47,7 +47,7 @@ def _get_target_arg(self, target: Target) -> str:
             if target.version == 'v4':
                 return f'--ip-v4-target={target.address}'
             elif target.version == 'v6':
-                return f'--ip-v4-target={target.address}'
+                return f'--ip-v6-target={target.address}'
```
Comment (deadly-panda) [inline agent/tsunami/tsunami.py:50]: nice catch!

**Review by deadly-panda — CHANGES_REQUESTED** [review]

**Review by 3asm — APPROVED** [review]

File: agent/tsunami_agent.py [hunk 9]
```diff
@@ -46,6 +51,35 @@ def _check_asset_was_added(self, targets: tsunami.Target) -> bool:
...
+            if url.port is not None:
+                metadata_type = agent_report_vulnerability_mixin.MetadataType.PORT
+                metadata_value = str(url.port)
+                metadata = [
+                    agent_report_vulnerability_mixin.VulnerabilityLocationMetadata(type=metadata_type,
+                                                                                   value=metadata_value)
+                ]
+            if url.scheme != '':
+                asset = link_asset.Link(url=target.domain, method='GET')
+            else:
```
Comment (3asm) [inline agent/tsunami_agent.py:78]: asset should be domain, not link.

Comment (3asm) [reply agent/tsunami_agent.py:78]: And metadata should have the URL.

**Review by 3asm — CHANGES_REQUESTED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**